spec/components/Web3API_Client.md

```diff
@@ -0,0 +1,157 @@
+# Web3API Client
+
+Web3API client is intended for application developers that can use the client as a single library in their (d)apps to interact with various decentralized services using Web3API's packages.
+
+Currently, Web3API client is available in the following languages:
+ - JavaScript (TypeScript)
+
+## Client interface
+The following snippet contains the interface that client should implement.
+
+```
+interface Web3APIClient {
+  redirects(): UriRedirect[];
+  query(options: QueryApiOptions): QueryApiResult;
+}
+
+type UriRedirect {
+ // Redirect from this URI or set of URIs specified by the regex
+ from: Uri | RegExp;
+ // Destination URI or plugin that will handle invocations
+ to: Uri | (() => Plugin)
+}
+
+type QueryApiOptions {
+  // Target URI that is being queried
+  uri: Uri;
+  // GraphQL query
+  query: string | QueryDocument;
+  // Optional variables that are part of the GraphQL query
+  variables?: Record<string, any>;
+}
+
+type QueryApiResult {
+ /**
+   * Query result data. The type of this value is a named map,
+   * where the key is the method's name, and value is the [[InvokeApiResult]]'s data.
+   * This is done to allow for parallel invocations within a
+   * single query document. In case of method name collisions,
+   * a postfix of `_0` will be applied, where 0 will be incremented for
+   * each occurrence. If undefined, it means something went wrong.
+ */
+  data?: [[InvokeApiResult]];
```

Comment:
ain't this suppose to be map or tuple?
Not entirely sure how to hanle "any" value in other languages. I wonder if there are some restrictions imposed by msg pack on type of result.